cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,407 @@
+        result = protocols.resolve_parameters(result, self.param_resolver)
+
+        return list(result.all_operations()) * abs(self.repetitions)
+
+    # Methods for string representation of the operation.
+
+    def __repr__(self):
+        args = [f'circuit={self.circuit!r}']
+        if self.repetitions != 1:
+            args.append(f'repetitions={self.repetitions}')
+        if self.qubit_map:
+            args.append(f'qubit_map={proper_repr(self.qubit_map)}')
+        if self.measurement_key_map:
+            args.append(f'measurement_key_map={proper_repr(self.measurement_key_map)}')
+        if self.param_resolver:
+            args.append(f'param_resolver={proper_repr(self.param_resolver)}')
+        argstr = ",\n".join(args)  # no backslashes allowed in fstring braces
+        return f'cirq.CircuitOperation({argstr})'
+
+    def __str__(self):
+        # TODO: support out-of-line subcircuit definition in string format.
+        header = self.circuit.serialization_key() + ':'
+        msg_lines = str(self.circuit).split('\n')
+        msg_width = max([len(header) - 4] + [len(line) for line in msg_lines])
+        circuit_msg = '\n'.join(
+            ['[ {line:<{width}} ]'.format(line=line, width=msg_width) for line in msg_lines]
+        )
+        args = []
+
+        def dict_str(d: Dict) -> str:
+            pairs = [f'{k}: {v}' for k, v in sorted(d.items())]
+            return '{' + ', '.join(pairs) + '}'
+
+        if self.qubit_map:
+            args.append(f'qubit_map={dict_str(self.qubit_map)}')
+        if self.measurement_key_map:
+            args.append(f'key_map={dict_str(self.measurement_key_map)}')
+        if self.param_resolver:
+            args.append(f'params={self.param_resolver.param_dict}')
+        if self.repetitions != 1:
+            args.append(f'loops={self.repetitions}')
+        if not args:
+            return f'{header}\n{circuit_msg}'
+        return f'{header}\n{circuit_msg}({", ".join(args)})'
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': 'CircuitOperation',
+            'circuit': self.circuit,
+            'repetitions': self.repetitions,
+            # JSON requires mappings to have keys of basic types.
+            # Pairs must be sorted to ensure consistent serialization.
+            'qubit_map': sorted(self.qubit_map.items()),
+            'measurement_key_map': self.measurement_key_map,
+            'param_resolver': self.param_resolver,
+        }
+
+    @classmethod
+    def _from_json_dict_(
+        cls, circuit, repetitions, qubit_map, measurement_key_map, param_resolver, **kwargs
+    ):
+        result = (
+            cls(circuit)
+            .with_qubit_mapping(dict(qubit_map))
+            .with_measurement_key_mapping(measurement_key_map)
+            .with_params(param_resolver)
+        )
+        if repetitions != 1:
+            return result.repeat(repetitions)
```

Comment:
Done. Added a note to parallel the notes in GateOperation: this is only legal because `CircuitOperation` is (intended to be) immutable. Otherwise we'd need to adhere to the "returns a copy" policy more rigorously.